### PR TITLE
Adding LoadScalable class

### PR DESCRIPTION
**Pull request by WalAmeni**
Deprecated listGenerators should be replaced with filterInjections

### CODE REVIEW

File: action/action-util/src/test/java/com/powsybl/action/util/LoadScalableTest.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    @Test
+    public void testMaximumlValue() {
+        assertEquals(Double.POSITIVE_INFINITY, l1.maximumValue(network), 1e-3);
+    }
+
+    @Test
+    public void testLoadScale() {
+        Load load = network.getLoad("l1");
+        assertEquals(100, load.getP0(), 1e-3);
+        assertEquals(20, l1.scale(network, 20), 1e-3);
```
Comment (mathbagu) [inline action/action-util/src/test/java/com/powsybl/action/util/LoadScalableTest.java]: This is not true: if initial P is equal to 100, if I want to scale to 20MW, I expect 120MW. Regarding your test, you did a -80MW scaling.

File: action/action-util/src/main/java/com/powsybl/action/util/GeneratorScalable.java [hunk 2]
```diff
@@ -47,48 +65,98 @@ public void reset(Network n) {
...
+        if (oldTargetP < this.minimumValue(n) || oldTargetP > this.maximumValue(n)) {
+            throw new PowsyblException("Error scaling GeneratorScalable " + id +
+                    " : Initial P is not in the range [Pmin, Pmax]");
+        }
+
+        double availableUp = maximumValue(n, ScalingPowerConvention.GENERATOR) - oldTargetP;
+        double availableDown = oldTargetP - minimumValue(n, ScalingPowerConvention.GENERATOR);
+
+
+        if (ScalingPowerConvention.GENERATOR.equals(scalingConvention)) {
```
Comment (mathbagu) [inline action/action-util/src/main/java/com/powsybl/action/util/GeneratorScalable.java]: Use an equality

File: action/action-util/src/main/java/com/powsybl/action/util/LoadScalable.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+ */
+class LoadScalable extends AbstractScalable {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(LoadScalable.class);
+
+    private final String id;
+
+    private final double maxValue;
+
+    private final double minValue;
```
Comment (mathbagu) [inline action/action-util/src/main/java/com/powsybl/action/util/LoadScalable.java]: Same remark about the minValue and maxValue order

File: action/action-util/src/main/java/com/powsybl/action/util/LoadScalable.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    LoadScalable(String id, double maxValue) {
+        this(id, 0., maxValue);
+    }
+
+    LoadScalable(String id, double minValue, double maxValue) {
+        this.id = Objects.requireNonNull(id);
+        if (maxValue < minValue) {
+            throw new PowsyblException("Error creating LoadScalable " + id
+                    + " : maxValue should be bigger than minValue");
+        } else {
```
Comment (mathbagu) [inline action/action-util/src/main/java/com/powsybl/action/util/LoadScalable.java]: Remove this else

File: action/action-util/src/main/java/com/powsybl/action/util/LoadScalable.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+        Load l = n.getLoad(id);
+        if (l != null) {
+            l.setP0(0);
+        }
+    }
+
+    /**
+     * Set a maximum value of active power for the LoadScalable
+     * <p> by default Double.MAX_VALUE
```
Comment (mathbagu) [inline action/action-util/src/main/java/com/powsybl/action/util/LoadScalable.java]: Does this <p> really needed?

File: action/action-util/src/main/java/com/powsybl/action/util/LoadScalable.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+     */
+    @Override
+    public double maximumValue(Network n, ScalingPowerConvention scalingConvention) {
+        return ScalingPowerConvention.LOAD.equals(scalingConvention) ? maxValue : -minValue;
+    }
+
+
+    /**
+     * Set a minimum value of active power for the LoadScalable
+     * <p> by default zero
```
Comment (mathbagu) [inline action/action-util/src/main/java/com/powsybl/action/util/LoadScalable.java]: Same remark here

File: action/action-util/src/main/java/com/powsybl/action/util/LoadScalable.java [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    @Override
+    public void filterInjections(Network n, List<Injection> injections, List<String> notFoundInjections) {
+        Objects.requireNonNull(n);
+        Objects.requireNonNull(injections);
+
+        Load load = n.getLoad(id);
+        if (load != null) {
+            injections.add(load);
+        } else {
+            if (notFoundInjections != null) {
```
Comment (mathbagu) [inline action/action-util/src/main/java/com/powsybl/action/util/LoadScalable.java]: Merge the `else` and the `if` statements

File: action/action-util/src/main/java/com/powsybl/action/util/LoadScalable.java [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+        }
+    }
+
+
+    /**
+     * @param n                 network
+     * @param asked             value asked to adjust the scalable active power
+     * @param scalingConvention
+     * If scalingConvention is LOAD, the load active power increases for positive "asked" and decreases inversely
+     * If scalingConvention is GENERATOR, the load active power decreases for positive "asked" and increases inversely
```
Comment (mathbagu) [inline action/action-util/src/main/java/com/powsybl/action/util/LoadScalable.java]: Move these comments before params

File: action/action-util/src/main/java/com/powsybl/action/util/LoadScalable.java [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+     * @return actual value of adjusted active power in accordance with the scaling convention
+     */
+    @Override
+    public double scale(Network n, double asked, ScalingPowerConvention scalingConvention) {
```
Comment (mathbagu) [inline action/action-util/src/main/java/com/powsybl/action/util/LoadScalable.java]: Assert that scalingConvention is not null.

File: action/action-util/src/main/java/com/powsybl/action/util/AbstractCompoundScalable.java [hunk 10]
```diff
@@ -43,22 +43,46 @@ public void reset(Network n) {
 
     @Override
     public double maximumValue(Network n) {
+        return maximumValue(n, ScalingPowerConvention.GENERATOR);
+    }
+
+    @Override
+    public double maximumValue(Network n, ScalingPowerConvention powerConvention) {
```
Comment (mathbagu) [inline action/action-util/src/main/java/com/powsybl/action/util/AbstractCompoundScalable.java]: Rename to scalingConvention to be consistent

File: action/action-util/src/main/java/com/powsybl/action/util/AbstractCompoundScalable.java [hunk 11]
```diff
@@ -43,22 +43,46 @@ public void reset(Network n) {
...
     }
 
     @Override
-    public void listGenerators(Network n, List<Generator> generators, List<String> notFoundGenerators) {
+    public double minimumValue(Network n) {
+        return minimumValue(n, ScalingPowerConvention.GENERATOR);
+    }
+
+    @Override
+    public double minimumValue(Network n, ScalingPowerConvention powerConvention) {
```
Comment (mathbagu) [inline action/action-util/src/main/java/com/powsybl/action/util/AbstractCompoundScalable.java]: Same remark here

File: action/action-util/src/main/java/com/powsybl/action/util/LoadScalable.java [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+        }
+
+        Terminal t = l.getTerminal();
+        if (!t.isConnected()) {
+            t.connect();
+            LOGGER.info("Connecting {}", l.getId());
+        }
+
+        double oldP0 = l.getP0();
+        if (oldP0 < this.minimumValue(n, ScalingPowerConvention.LOAD) || oldP0 > this.maximumValue(n, ScalingPowerConvention.LOAD)) {
```
Comment (mathbagu) [inline action/action-util/src/main/java/com/powsybl/action/util/LoadScalable.java]: remove the (useless ?) `this.`

add a static import of the conventions

File: action/action-util/src/main/java/com/powsybl/action/util/ProportionalScalable.java [hunk 13]
```diff
@@ -37,15 +37,16 @@ private static void checkPercentages(List<Float> percentages, List<Scalable> sca
         this.percentages = Objects.requireNonNull(percentages);
     }
 
+
     @Override
-    public double scale(Network n, double asked) {
+    public double scale(Network n, double asked, ScalingPowerConvention scalingConvention) {
         Objects.requireNonNull(n);
```
Comment (mathbagu) [inline action/action-util/src/main/java/com/powsybl/action/util/ProportionalScalable.java]: Assert that scalingConvention is not null

File: action/action-util/src/main/java/com/powsybl/action/util/StackScalable.java [hunk 14]
```diff
@@ -25,15 +25,16 @@
...
-    public double scale(Network n, double asked) {
+    public double scale(Network n, double asked, ScalingPowerConvention scalingConvention) {
         Objects.requireNonNull(n);
 
         double done = 0;
         double remaining = asked;
         for (Scalable scalable : scalables) {
-            if (remaining > 0) {
-                double v = scalable.scale(n, remaining);
+            if (Math.abs(remaining) > 0) {
```
Comment (mathbagu) [inline action/action-util/src/main/java/com/powsybl/action/util/StackScalable.java]: is it equivalent to `remaining != 0` ?

Comment (murgeyseb) [reply action/action-util/src/main/java/com/powsybl/action/util/StackScalable.java]: Following this example (http://www.java67.com/2015/09/float-and-double-value-comparison-in-java-use-relational.html) I think we should use an EPSILON threshold don't you think ?

Comment (mathbagu) [reply action/action-util/src/main/java/com/powsybl/action/util/StackScalable.java]: Sure! It’s better with an epsilon

**Comment by murgeyseb** [comment]
> I'm not sure the code works properly if the asked value is negative. I don't find a unit test with that cases, specially when the convention is not the default one.
> 
> Could you add more tests?

I see that there are tests for each convention in each sign (positive and negative).

File: action/action-util/src/main/java/com/powsybl/action/util/AbstractScalable.java [hunk 15]
```diff
@@ -21,14 +23,71 @@ protected AbstractScalable() {
...
+    @Deprecated
     public List<Generator> listGenerators(Network n) {
         return listGenerators(n, null);
     }
+
+    /**
+     * @deprecated listGenerators should be replaced by filterInjections
+     */
+    @Deprecated
+    public void listGenerators(Network network, List<Generator> generators, List<String> notFoundGenerators) {
```
Comment (geofjamg) [inline action/action-util/src/main/java/com/powsybl/action/util/AbstractScalable.java]: Could you check this method is correctly implemented? It seems to me that notFoundgenerators can contain loads.

File: action/action-util/src/main/java/com/powsybl/action/util/GeneratorScalable.java [hunk 16]
```diff
@@ -48,47 +63,89 @@ public void reset(Network n) {
...
 
         Generator g = n.getGenerator(id);
         if (g != null) {
-            generators.add(g);
+            return scalingConvention == GENERATOR ? Math.max(g.getMinP(), minValue) : -Math.min(g.getMaxP(), maxValue);
         } else {
-            if (notFoundGenerators != null) {
-                notFoundGenerators.add(id);
-            }
+            return 0;
```
Comment (geofjamg) [inline action/action-util/src/main/java/com/powsybl/action/util/GeneratorScalable.java]: -Double.MAX_VALUE ?

File: action/action-util/src/main/java/com/powsybl/action/util/GeneratorScalable.java [hunk 17]
```diff
@@ -48,47 +63,89 @@ public void reset(Network n) {
...
+    public double maximumValue(Network n, ScalingConvention scalingConvention) {
         Objects.requireNonNull(n);
+        Objects.requireNonNull(scalingConvention);
 
         Generator g = n.getGenerator(id);
-        return g != null ? g.getMaxP() : 0;
+        if (g != null) {
+            return scalingConvention == GENERATOR ? Math.min(g.getMaxP(), maxValue) : -Math.max(g.getMinP(), minValue);
+        } else {
+            return 0;
```
Comment (geofjamg) [inline action/action-util/src/main/java/com/powsybl/action/util/GeneratorScalable.java]: Double.MAX_VALUE?

File: action/action-util/src/main/java/com/powsybl/action/util/AbstractScalable.java (hunk 15 above)
Comment (murgeyseb) [reply action/action-util/src/main/java/com/powsybl/action/util/AbstractScalable.java]: In fact, not found generators may contain load IDs. It is a way to inform the user that it cannot find a generator that has as ID the ID of another kind of injection.
That's how we designed it, but as you see it is put as deprecated, so it is supposed to be removed at the end. You would prefer that we check it does not exist at all in the network ? Log an error or throw an exception if it is an object of other type ?

File: action/action-util/src/main/java/com/powsybl/action/util/GeneratorScalable.java (hunk 16 above)
Comment (murgeyseb) [reply action/action-util/src/main/java/com/powsybl/action/util/GeneratorScalable.java]: No, if the generator exist, you can shift 0 on the Scalable, so minValue = maxValue = initialValue = 0

File: action/action-util/src/main/java/com/powsybl/action/util/GeneratorScalable.java (hunk 17 above)
Comment (murgeyseb) [reply action/action-util/src/main/java/com/powsybl/action/util/GeneratorScalable.java]: Same as before

File: action/action-util/src/main/java/com/powsybl/action/util/LoadScalable.java [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+        Load l = n.getLoad(id);
+
+        double done = 0;
+        if (l == null) {
+            LOGGER.warn("Load {} not found", id);
+            return done;
+        }
+
+        Terminal t = l.getTerminal();
+        if (!t.isConnected()) {
```
[the remark this answers is not in the document]
Comment (murgeyseb) [reply action/action-util/src/main/java/com/powsybl/action/util/LoadScalable.java]: We kept the initial behavior here. Perhaps allowing to overload this behavior may be part of another feature request ?

File: action/action-util/src/main/java/com/powsybl/action/util/GeneratorScalable.java (hunk 16 above)
Comment (geofjamg) [reply action/action-util/src/main/java/com/powsybl/action/util/GeneratorScalable.java]: OK

File: action/action-util/src/main/java/com/powsybl/action/util/GeneratorScalable.java (hunk 17 above)
Comment (geofjamg) [reply action/action-util/src/main/java/com/powsybl/action/util/GeneratorScalable.java]: OK

File: action/action-util/src/main/java/com/powsybl/action/util/AbstractScalable.java (hunk 15 above)
Comment (geofjamg) [reply action/action-util/src/main/java/com/powsybl/action/util/AbstractScalable.java]: As this method will be removed soon and  is not expected to be used mixed with loads, we can let t as it is.

File: action/action-util/src/main/java/com/powsybl/action/util/LoadScalable.java [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+
+import java.util.List;
+import java.util.Objects;
+
+import static com.powsybl.action.util.Scalable.ScalingConvention.*;
+
+/**
+ * @author Ameni Walha <ameni.walha at rte-france.com>
+ */
+class LoadScalable extends AbstractScalable {
```
[the remark this answers is not in the document]
Comment (murgeyseb) [reply action/action-util/src/main/java/com/powsybl/action/util/LoadScalable.java]: Done

File: action/action-util/src/main/java/com/powsybl/action/util/LoadScalable.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (geofjamg) [reply action/action-util/src/main/java/com/powsybl/action/util/LoadScalable.java]: OK I created the issue #716

**Pull request merged**